clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package etcdinit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/clustermesh-apiserver/etcdinit/defaults"
+	einit "github.com/cilium/cilium/pkg/etcd/init"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/spf13/cobra"
+	client "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	log                     = logging.DefaultLogger.WithField(logfields.LogSubsys, "etcdinit")
+	etcdBinaryLocation      = "/usr/bin/etcd"
+	etcdDataDir             = ""
+	etcdClusterName         = ""
+	etcdInitialClusterToken = ""
+	debug                   = false
+)
+
+func NewCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "etcdinit",
+		Short: "Initialise an etcd data directory for use by the etcd sidecar of clustermesh-apiserver",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
+	rootCmd.Flags().StringVar(&etcdInitialClusterToken, "initial-cluster-token", defaults.EtcdInitialClusterToken, "etcd initial cluster token")
+	rootCmd.Flags().BoolVar(&debug, "debug", false, "Debug log output")
+	return rootCmd
+}
+
+func InitEtcdLocal() error {
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancelFn()
+
+	if debug {
+		logging.SetLogLevelToDebug()
+		log.Debug("Debug logging enabled")
+	}
+
+	log := log.
+		WithField("etcdBinaryLocation", etcdBinaryLocation).
+		WithField("etcdDataDir", etcdDataDir).
+		WithField("etcdClusterName", etcdClusterName).
+		WithField("etcdInitialClusterToken", etcdInitialClusterToken)
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("path", d.Name()).Debug("Removing file/directory in data dir")
+		os.RemoveAll(path.Join(etcdDataDir, d.Name()))
+	}
+
+	log.Info("Starting localhost-only etcd process")
+	// Specify the full path to the etcd binary to avoid any PATH search binary replacement nonsense
+	endpoint := "http://127.0.0.1:2379"
+	etcdCmd := exec.CommandContext(ctx, etcdBinaryLocation,
+		fmt.Sprintf("--data-dir=%s", etcdDataDir),
+		fmt.Sprintf("--name=%s", etcdClusterName),
+		// This probably won't work on an IPv6 only cluster
```

Comment:
:thinking: Wondering the same honestly, but it was previously working, so likely etcd has some logic to handle this case (as IIRC it is also the default value). In any case we'll see from the E2E tests, as at least one of them creates an IPv6-only cluster.

Reply:
Yeah, this is what the existing code did, so it's not going to be any **worse**.

Reply:
Fair enough. I considered the case in which nodes use IPv6 addresses only (which is what is tested there, and that used to cause issues in the past to the main etcd container), but not the possibility of completely disabling IPv4 networking (and thus also the 127.0.0.1 loopback address) at the kernel level.

I guess that using `http://localhost:2379` as endpoint could be better than (as it is also the default value for these settings).

Reply:
Done